8323220: Reassociate loop invariants involved in Cmps and Add/Subs

caojoshua opened the pull request:
```
// inv1 == (x + inv2) => ( inv1 - inv2 ) == x
// inv1 == (x - inv2) => ( inv1 + inv2 ) == x
// inv1 == (inv2 - x) => (-inv1 + inv2 ) == x
```

For example,

```
fn(inv1, inv2)
  while(...)
      x = foobar()
      if inv1 == x + inv2
           blackhole()
```

We can transform this into

```
fn(inv1, inv2)
  t = inv1 - inv2
  while(...)
      x = foobar()
      if t == x
           blackhole()
```

Here is an example: https://github.com/openjdk/jdk/blob/b78896b9aafcb15f453eaed6e154a5461581407b/src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java#L910. LHS `1` and RHS `pos` are both loop invariant

Passes tier1 locally on Linux machine. Passes GHA on my fork.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8323220](https://bugs.openjdk.org/browse/JDK-8323220): Reassociate loop invariants involved in Cmps and Add/Subs (**Enhancement** - P5)


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Xin Liu](https://openjdk.org/census#xliu) (@navyxliu - Committer) ⚠️ Review applies to [5ea7a53a](https://git.openjdk.org/jdk/pull/17375/files/5ea7a53ae666472f1bf24f86e4cd9f169e10ff7b)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17375/head:pull/17375` \
`$ git checkout pull/17375`

Update a local copy of the PR: \
`$ git checkout pull/17375` \
`$ git pull https://git.openjdk.org/jdk.git pull/17375/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17375`

View PR using the GUI difftool: \
`$ git pr show -t 17375`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17375.diff">https://git.openjdk.org/jdk/pull/17375.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17375#issuecomment-1887670114)

caojoshua commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,221 @@
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(failOn = {IRNode.SUB_I})
+    public void leDontReassociate(int inv1, int inv2) {
```
I added DontReassociate tests for `le`, `gt`, and `ge`. For `lt`, C2 generates a second `SUB_I` as part of other transformations.

IR matching for ADD/SUB is pretty hard in general. They commonly are created as part of other transformations. Any suggestions on how I can test this better is appreciated.

eme64 replied:
You could always use a simple regex with the linenumber I guess. But that is a bit nasty too.

caojoshua replied:
Yeah, it would work for this patch. But people working on future unrelated changes may have to change the line number. Seems more pain than its worth.

eme64 replied:
Another suboptimal idea: you wrap the add / sub in a method, and then ensure that this method is inlined. It might still keep the annotation of being part of that inner method, and you could use regex to check for it.

Or maybe we could also have some sort of relative line offset mechanism in the IR framework, that allows you to specify that you want something that is let's say 7 lines down from the IR rule.

caojoshua replied:
Some sort of pattern matcher could work. It would be able nice to match something like `a ADD_I b CMP_LT c`. In java this could look something like 

```
@IR(counts = {IRNode.CMP_LT[IRNode.ANY, IRNode.SUB_I, IRNode.ANY], "1"}
```

The arguments in the `[]` are the inputs. `IRNode.ANY` matches any node. (The zero'th node is ANY because its the region node).

Anyway, I think a `lt` test is not super-required for the coverage for this PR. The current machinery does not provide a convenient way to test it. I'd prefer to avoid something hacky. I think this work can be done separately.

eme64 replied:
I agree with you there, don't do anything hacky here.
But yes, I've also been wondering what kind of improvements to the IR framework would help us to do these sorts of graph-matching verifications.

caojoshua replied:
Created https://bugs.openjdk.org/browse/JDK-8324226

eme64 commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -263,6 +263,25 @@ int IdealLoopTree::find_invariant(Node* n, PhaseIdealLoop *phase) {
+    return false;
+  }
+  for (DUIterator i = n->outs(); n->has_out(i); i++) {
+    BoolNode *boolOut = n->out(i)->isa_Bool();
```
```suggestion
    BoolNode* boolOut = n->out(i)->isa_Bool();
```

eme64 commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -305,22 +325,29 @@ bool IdealLoopTree::is_associative(Node* n, Node* base) {
+                  (n1->is_Cmp() && n2->is_Add());
+  bool neg_inv1 = (n1->is_Sub() && !n1->is_Cmp() && inv1_idx == 2) ||
+                  (n1->is_Cmp() && inv2_idx == 1 && n2->is_Sub());
+  if (n1->is_Sub() && !n1->is_Cmp() && inv1_idx == 1) {
```
Would you mind adding some comments for this logic?

caojoshua replied:
I added a little comment block. Not sure how useful it is. I agree the code is hard to follow. Before all the changes for `Cmp`, I was able to make sense of it by following the comments at the top of the function.

eme64 replied:
Hmm, how about using these variables?
```
bool n1_is_add = n1->is_Add();
bool n1_is_sub = n1->is_Sub() && !n1->is_Cmp();
bool n1_is_cmp = n1->is_Cmp();
```
Then you can just comment this:
`// Determine whether x, inv1, or inv2 should be negative in the transformed expression`

eme64 reviewed (changes requested):
Nice idea, thanks for the work!

Can you also create a regression test that has edge-case values and random values, to check that the correctness ok? This would also experimentally rule out overflow issues.

eme64 reviewed (changes requested):
Thanks for the update, looks better already!
I'm still waiting for the test with random/edge-case values, and then I can submit this for testing :)

caojoshua commented:
> Thanks for the update, looks better already! I'm still waiting for the test with random/edge-case values, and then I can submit this for testing :)

I converted two of the "dont-associate" tests to edge case values, and a couple of the positive tests to have random arguments.

Changing other tests end up with the issue where there are more add/sub nodes that make it hard to match IR.

navyxliu commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -346,6 +379,9 @@ Node* IdealLoopTree::reassociate_add_sub(Node* n1, int inv1_idx, int inv2_idx, P
       inv = new AddINode(n_inv1, inv2);
     }
     phase->register_new_node(inv, phase->get_early_ctrl(inv));
+    if (n1_is_cmp) {
```
CmpNode is subclass of SubNode. if n is CmpI/L, n->is_Sub() is also true. 
can we use the old logic for your new comparison expressions? 

yes, we still need to check if n1 is CmpNode or SubNode here.

navyxliu replied:
I think CmpNode has the same re-association rule as SubNode. 
take your own example, 
>  inv1 == (x - inv2) => ( inv1 + inv2 ) == x

Cmp(inv1, (x-inv2)) => Eq(0, Sub(inv1, (x-inv2)) => Eq(0, Sub(inv1+inv2, x)) => Cmp(inv1+inv2, x)

Originally, n1->is_Sub() covers both CmpNode and SubNode. I don't you need to split them it into 2 cases. 

I think you only to check if n1 Cmp or Sub when you are going to return Cmp/SubNode

caojoshua replied:
The rules are not always the same. For example

> inv1 - (inv2 - x)     => (inv1 - inv2) + x
> inv1  == (inv2 - x) => (-inv1 + inv2) == x

The signs of inv1 and inv2 are flipped. I don't think we can fold these into a single conditional

navyxliu replied:
okay. I see you mean.

navyxliu commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -305,22 +325,35 @@ bool IdealLoopTree::is_associative(Node* n, Node* base) {
   }
 
-  bool is_int = n1->bottom_type()->isa_int() != nullptr;
+  bool is_int = n2->bottom_type()->isa_int() != nullptr;
```
I guess you change to n2 because n1 may be CmpI/L. 
I think it still works because TypeInt::CC is still isa_int(). or we add a comment to make it more clear?

navyxliu replied:
Sorry, I think you're right. we have to use 'n2' here. 
we need to distinct CmpI and CmpL. TypeInt::CC can't.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public void notEqualsInvariantSubVariantLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv2 - inv1 != i`
+            if (inv1 - i != inv2) {
+                blackhole();
+            }
+        }
+    }
```
Why could you not have random arguments here?

caojoshua replied:
[notEqualsInvariantSubVariantLong.txt](https://github.com/openjdk/jdk/files/14015703/notEqualsInvariantSubVariantLong.txt)

Attaching the IdealGraphVisualizer file (renamed to `.txt` cause GitHub wants that).

When we use `Argument.NUMBER_42`, we can avoid generating traps. When `i == 0` we enter the if statement, otherwise we exit.

If we use random numbers, we most likely never enter the if statement so we can generate a trap there. In this case, the trap block as an extra `SubL` for `inv1 - i`.

navyxliu replied:
As long as you have inv1 == inv2, your comparison will have a different result when i == 0. 

Can you randomize just one value and use them for both inv1 and inv2?

eme64 replied:
@caojoshua I see. That is a sad limitation. You could now add a `@Run` statement, but that is overkill as well. I hope to improve the IR framework to make it a bit easier to pass better arguments soon. For now I think we could just leave it with constants, for simplicity.

caojoshua replied:
> Can you randomize just one value and use them for both inv1 and inv2?

Yeah we can just take a single argument `inv1`. I'd prefer having two arguments and leaving them constant.

eme64 reviewed (changes requested):
Thanks for adding some random arguments!

One more nice thing might be result verification, i.e. verify that the generated code actually returns the correct results, for example the expected `i`.

navyxliu reviewed (approved):
LTGM. I am not a reviewer. 
just one suggestion. Should we also update the comments of 'reassociate' ?  this line:
> For add/sub expressions: see "reassociate_add_sub"

eme64 commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -305,22 +325,35 @@ bool IdealLoopTree::is_associative(Node* n, Node* base) {
+// inv1 == (x - inv2) => ( inv1 + inv2 ) == x
+// inv1 == (inv2 - x) => (-inv1 + inv2 ) == x
+//
+Node* IdealLoopTree::reassociate_add_sub_cmp(Node* n1, int inv1_idx, int inv2_idx, PhaseIdealLoop* phase) {
```
as @navyxliu said, you should also update any comments with the old name. One thing I have spotted is this:
`//---------------------reassociate_add_sub------------------------`

eme64 commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -743,12 +743,14 @@ class IdealLoopTree : public ResourceObj {
   // Reassociate invariant binary expressions.
   Node* reassociate(Node* n1, PhaseIdealLoop *phase);
   // Reassociate invariant add and subtract expressions.
```
I guess you could mention cmp here too.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
```
Does the Amazon copyright header not have a year  associated with it?

eme64 replied:
I guess not, I see other files without a year. Still, a bit strange.

navyxliu replied:
Yes. Our open-source board advise us to do so.

eme64 reviewed (approved):
Tests pass, thanks for the updates.
I left a few optional comments.

One requiest I would still like to see:
You could improve the tests, by returning a value that indicates when the loop was exited, i.e. `return i`. In the future, I intend to verify the return values from test methods, and then we would have additional coverage for free ;)

caojoshua commented:
failure is due to mac infra issues and not related to test. @eme64 do the @check's look ok?

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,253 @@
+            throw new RuntimeException("Illegal reassociation");
+        }
+    }
+
```
```suggestion
```

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,253 @@
+    }
+
+    @Check(test = "geDontReassociate")
+    public void checkGeDontReassociate(int returnValue, TestInfo info) {
```
```suggestion
    public void checkGeDontReassociate(int returnValue) {
```

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,253 @@
+    }
+
+    @Check(test = "gtDontReassociate")
+    public void checkGtDontReassociate(int returnValue, TestInfo info) {
```
```suggestion
    public void checkGtDontReassociate(int returnValue) {
```

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,253 @@
+    }
+
+    @Check(test = "leDontReassociate")
+    public void checkLeDontReassociate(int returnValue, TestInfo info) {
```
```suggestion
    public void checkLeDontReassociate(int returnValue) {
```

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,253 @@
+                blackhole();
+            }
+        }
+    }
```
I suggest you add a test that first iterates some number of iterations, and then returns inside the if-statement with return value i. Then you check if that value is as expected.
Because with all the tests that you have here, I'm not sure if really all branches are always taken, or if they may lead to uncommon traps, and hence leave some things uncovered by testing.

caojoshua replied:
@eme64 can you take a look at the updates?

eme64 replied:
I don't think you have added what I requested. Do you have any test that does the reassociation, and where you verify that the computed values are correct?

caojoshua replied:
OK thanks I think I misread. I now have some correctness tests for tests with and without reassociation. 

We can't add correctness tests for tests with random inputs because we do not know how many iterations there will be. There are two tests with reassociation `equalsVariantSubInvarint(Int|Long)` that have a drastically different graph with the correctness change that messes up the IR matching, so I opted out of changes those as well.

eme64 replied:
In fact, that would probably be smart. Otherwise you only have result verification for constant values. And who knows, there the code might constant-fold in a way that we don't expect.

caojoshua replied:
I am with you that I am not 100% happy with the tests. I do not mind waiting for that PR to get integrated.

If I understand correctly, that PR is more for convenience of setting up arguments. How will that help testing correctness for random inputs?

caojoshua replied:
I added correctness tests for two of the tests with random inputs. For the rest, it messes up the IR matching. This goes back to the fragility of matching add/sub nodes.

eme64 reviewed (changes requested):
I thought the `TestInfo info` was optional, so if you don't use it, then don't put it in the arguments ;)

eme64 commented:
https://github.com/openjdk/jdk/pull/17557
Is integrated. You need to merge with master now, and inspect all usages of `@Arguments`.

I imagine it like this:
```
public class InvariantCodeMotionReassociateCmp {
    int inv1Ref, inv2Ref, invDiff; // have fields to use in @Setup and @Check

    @Setup
    public Object[] setupArguments() {
        inv1Ref = RANDOM.nextInt();
        ... // setup your reference values in a smart way, so that half of the time the test returns -1, and half the time it returns i.
        return new Object[]{ inv1Ref, inv2Ref };
    }
    
    @Test
    @Arguments(setup = "setupArguments")
    @IR(counts = {IRNode.SUB_I, "1"})
    public int[] equalsAddInt(int inv1, int inv2) {
        for (int i = 0; i < size; ++i) {
            // Reassociate to `inv2 - inv1 == i`
            if (inv1 + i == inv2) {
                return i;
            }
        }
        return -1;
    }
    
    @Check(test = "equalsAddInt")
    public void checkEqualsAddInt(int returnValue) {
        // now check if the returnValue is as expected, given the field values
    }
```
This should also have fewer add/sub nodes in the test method, I think.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,315 @@
+    public void checkEqualsAddLong(long[] returnValue) {
+        if (returnValue[0] != size && returnValue[0] + returnValue[1] != returnValue[2]) {
+            throw new RuntimeException("Illegal reassociation: i=" + returnValue[0] + ", inv1=" + returnValue[1]
+                    + ", inv2=" + returnValue[2]);
```
Sadly, you cannot reuse `@Check` methods for multiple `@Test` methods. But you can have a utility method that you invoke from multiple `@Check` methods. That would reduce the code-duplication.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,315 @@
+ * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
+ */
+public class InvariantCodeMotionReassociateCmp {
+    private static final int size = 500;
```
You could even randomize size in a `@Setup` method, and pass it as another argument.

caojoshua replied:
If I have a random size input, the loop is actually not a `CountedLoop`, and this transformation only applies to `CountedLoop`. I don't know all the details of `CountedLoop`, but this feels like a miss. Maybe `CountedLoop` can only handle loops with multiple exits of the limit is a constant.

eme64 replied:
Ha, yes, this would be a heavy miss. Can you post a small `Test.java` where such a loop does not get turned into a `CountedLoop`? I'd like to investigate.

eme64 replied:
Might it be that sometimes the size was too small, so that the loop was simply fully unrolled?

caojoshua replied:
My mistake. I was reading from a class member variable `size`. Probably cannot be a counted loop because the `size` can change by iteration.

Should be fine if I just pass a `size` variable.

caojoshua replied:
If I pass `size` variable, I get some weird matching patterns after strip mining transformations. I avoid this by moving `blackhole()` to before the if condition.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,315 @@
+        for (; i < size; ++i) {
+            // Reassociate to `inv2 - inv1 == i`
+            if (inv1 + i == inv2) {
+                blackhole();
```
If you return instead of break, do you still need the `blackhole`?

caojoshua replied:
No we don't need it. But it would feel strange to have a loop with nothing in it other than comparisons of an induction variable.

eme64 commented:
@caojoshua sorry, I hope this is not too annoying. I think you are doing good work. We will get there :)

About the add/sub matching fragility: do you know where the other add/sub are coming from? I wonder if there is anything that we could improve in the IR framework to help with that in the future.

I hope that in the future we can drop the `@Check` methods completely, see https://bugs.openjdk.org/browse/JDK-8310533
But that will take a while, since we also plan to refactor the IR framework a little.

caojoshua commented:
@eme64 thank you for integrating `@setup`. We can create random inputs, but still have enough control to avoid traps. I've added the missing `lt` check, and correctness checks for all tests.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,422 @@
+            inv2 = inv1 + RANDOM.nextInt(size * 2) + 1;
+        }
+        return new Object[] { inv1, inv2 };
+    }
```
```suggestion
    }

```

eme64 replied:
Have a free newline between methods

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,422 @@
+
+    public void checkEq(int returnValue) {
+        int invDiff = inv2 - inv1;
+        if ((invDiff < size && returnValue != invDiff) || (invDiff
+                    >= size && returnValue != size)) {
```
```suggestion
        if ((invDiff <  size && returnValue != invDiff) ||
            (invDiff >= size && returnValue != size)) {
```

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,422 @@
+
+    public void checkNe(int returnValue) {
+        int invDiff = inv2 - inv1;
+        if ((invDiff != 0 && returnValue != 0) || (invDiff
+                    == 0 && returnValue != 1)) {
```
```suggestion
        if ((invDiff != 0 && returnValue != 0) ||
            (invDiff == 0 && returnValue != 1)) {
```

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,422 @@
+    @Setup
+    public Object[] setupNe() {
+        inv1 = RANDOM.nextInt();
+        if (RANDOM.nextInt() % 7 == 0) {
+            if (inv1 == 0) {
+                inv1 = 1;
+            }
+            // Setup inputs to be equals sometimes to avoid uncommon traps
+            inv2 = inv1;
```
You may want to use the `SetupInfo` instead, and make this deterministic. Just in the case where the randomness here fails in all cases - then you would not have any equal case.
Also: why can `inv1` not be zero?

caojoshua replied:
Sure, can look into `SetupInfo`.

`inv1` can be zero. I'll remove that block.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,422 @@
+    public int equalsVariantSubInvariantInt(int inv1, int inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv1 + inv2 == i`
```
```suggestion
            // Reassociate to `inv2 - inv1 == i`
```
Or otherwise you'd have to explain why you do the `@IR(counts = {IRNode.SUB_I, "1"})` here

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,422 @@
+    public int equalsVariantSubInvariantLong(long inv1, long inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv1 + inv2 == i`
+            if (i - inv2 == -inv1) {
```
```suggestion
            // Reassociate to `inv2 - inv1 == i`
            if (i - inv2 == -inv1) {
```

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,422 @@
+    public int notEqualsVariantSubInvariantInt(int inv1, int inv2) {
+        int i = 0;
+        for (; i < 500; ++i) {
+            // Reassociate to `inv1 + inv2 != i`
+            if (i - inv2 != -inv1) {
```
```suggestion
            // Reassociate to `inv1 - inv2 != i`
            if (i - inv2 != -inv1) {
```

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,422 @@
+    public int notEqualsVariantSubInvariantLong(long inv1, long inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv1 + inv2 != i`
+            if (i - inv2 != -inv1) {
```
```suggestion
            // Reassociate to `inv1 - inv2 != i`
            if (i - inv2 != -inv1) {
```

eme64 reviewed (changes requested):
Good work with the @Setup methods! Thanks for the work :)

caojoshua commented:
@eme64 can you take another look?

eme64 commented:
@caojoshua it looks really good now!
I'm running our internal testing again, will report back.

One more concern I just had: do we have tests for the pre-existing Add/Sub reassociations? Because you now touched the logic around there we should make sure there are at least correctness tests. IR tests are probably basically impossible because it is the same number of Add/Sub nodes before and after the optimization.

I'd like to have another Reviewer look over this as well, therefore:
/reviewers 2 reviewer

eme64 commented:
@caojoshua 

I also ran our internal testing and it looks ok. But of course that is only on tests that we have, and if the other reassociations are not tested, then that helps little ;)

> Not that I know of. A bunch of reassociation was added in https://github.com/openjdk/jdk/commit/23ed3a9e91ac57295d274fefdf6c0a322b1e87b7, which does not have any tests.

Could you please add a result verification test per case of pre-existing reassociation? Otherwise I'm afraid it is hard to be sure you did not break those cases.

caojoshua commented:
@eme64 I added a bunch of tests for add/sub reassociation. I made them IR tests since there are some cases where the IR matching is interesting. I added node matching to each for completeness. But I'd say more importantly they test correctness.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateAddSub.java:
```diff
@@ -0,0 +1,525 @@
+ * @bug 8323220
+ * @summary Test loop invariant code motion of add/sub through reassociation
+ * @library /test/lib /
+ * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateAddSub
```
```suggestion
 * @run main compiler.c2.loopopts.InvariantCodeMotionReassociateAddSub
```
Otherwise external flags are not passed in.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,412 @@
+ * @bug 8323220
+ * @summary Test loop invariant code motion for cmp nodes through reassociation
+ * @library /test/lib /
+ * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
```
```suggestion
 * @run main compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
```
Just like above

eme64 reviewed (changes requested):
Just two little details, then we should be ready!
Ping me once you updated it, and I will run our internal testing one more time before approving!

caojoshua commented:
@eme64 thanks for reviewing. changes are in.

eme64 commented:
Code looks good, running testing now... Ping me again in 2 days if I don't report back by then ;)

caojoshua commented:
@eme64 how does testing look?

eme64 reviewed (approved):
Testing is clean.
Thanks for the work, it now looks good to me :)
(Still, I'd like another Reviewer to approve it.)

chhagedorn commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -258,14 +258,33 @@ void IdealLoopTree::compute_profile_trip_cnt(PhaseIdealLoop *phase) {
   return 0;
 }
 
+//---------------------is_associative_cmp-------------------------
```
I usually remove these legacy headers when touching old code. I don't think they add any benefit nowadays.

eme64 replied:
agree

chhagedorn commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -258,14 +258,33 @@ void IdealLoopTree::compute_profile_trip_cnt(PhaseIdealLoop *phase) {
+    return false;
+  }
+  for (DUIterator i = n->outs(); n->has_out(i); i++) {
+    BoolNode* boolOut = n->out(i)->isa_Bool();
+    if (boolOut == nullptr || !(boolOut->_test._test == BoolTest::eq ||
+                                boolOut->_test._test == BoolTest::ne)) {
```
We should use underscores instead of camelCase.
```suggestion
    BoolNode* bool_out = n->out(i)->isa_Bool();
    if (bool_out == nullptr || !(bool_out->_test._test == BoolTest::eq ||
                                 bool_out->_test._test == BoolTest::ne)) {
```

eme64 replied:
Agree, this is the C++ style. CamelCase is Java style.

chhagedorn commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -288,11 +307,12 @@ bool IdealLoopTree::is_associative(Node* n, Node* base) {
         || op == Op_OrI  || op == Op_OrL
-        || op == Op_XorI || op == Op_XorL;
+        || op == Op_XorI || op == Op_XorL
+        || is_associative_cmp(n);
```
You should also update the comment on L304 which does not mention cmps.

chhagedorn commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -288,11 +307,12 @@ bool IdealLoopTree::is_associative(Node* n, Node* base) {
 }
 
-//---------------------reassociate_add_sub------------------------
+//-------------------reassociate_add_sub_cmp---------------------
```
Can also be removed.

chhagedorn commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -308,22 +328,35 @@ bool IdealLoopTree::is_associative(Node* n, Node* base) {
+// inv1 == (x + inv2) => ( inv1 - inv2 ) == x
+// inv1 == (x - inv2) => ( inv1 + inv2 ) == x
+// inv1 == (inv2 - x) => (-inv1 + inv2 ) == x
+//
```
```suggestion
```

chhagedorn commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -308,22 +328,35 @@ bool IdealLoopTree::is_associative(Node* n, Node* base) {
-  bool neg_inv2 = n2->is_Sub() && inv2_idx == 2;
-  bool neg_inv1 = n1->is_Sub() && inv1_idx == 2;
-  if (n1->is_Sub() && inv1_idx == 1) {
+  // Determine whether x, inv1, or inv2 should be negative in the transformed
+  // expression
+  bool neg_x = n2_is_sub && inv2_idx == 1;
+  bool neg_inv2 =
+      (n2_is_sub && !n1_is_cmp && inv2_idx == 2) || (n1_is_cmp && !n2_is_sub);
+  bool neg_inv1 =
+      (n1_is_sub && inv1_idx == 2) || (n1_is_cmp && inv2_idx == 1 && n2_is_sub);
```
I suggest to remove the line breaks since the lines won't be longer than the added asserts above.

chhagedorn commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -738,13 +738,15 @@ class IdealLoopTree : public ResourceObj {
-  // Reassociate invariant add and subtract expressions.
-  Node* reassociate_add_sub(Node* n1, int inv1_idx, int inv2_idx, PhaseIdealLoop *phase);
+  // Reassociate invariant add, subtract, and compare expressions.
+  Node* reassociate_add_sub_cmp(Node* n1, int inv1_idx, int inv2_idx, PhaseIdealLoop *phase);
```
```suggestion
  Node* reassociate_add_sub_cmp(Node* n1, int inv1_idx, int inv2_idx, PhaseIdealLoop* phase);
```

chhagedorn commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateAddSub.java:
```diff
@@ -0,0 +1,525 @@
+ * @bug 8323220
+ * @summary Test loop invariant code motion of add/sub through reassociation
+ * @library /test/lib /
+ * @run main compiler.c2.loopopts.InvariantCodeMotionReassociateAddSub
```
You should use `driver` since we do not want to stress the driver VM with additionally passed VM flags like `-Xcomp` when running IR tests.
```suggestion
 * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateAddSub
```

eme64 replied:
But if we put `driver` here, do the flags get passed to the IR framework test VM?

chhagedorn replied:
Yes, I think for IR tests, we should not stress the driver VM. The sole purpose of the driver VM is to start the actual test VM (by passing the `javaoptions` and `vmoptions`) and do IR matching afterward. I don't think there is much benefit if we try to also stress the driver VM each time in a normal IR test.

What we could do (and I'm not sure if we already do that) is to have a dedicated (or a set of) IR test that runs with `main` such that the driver VM is also stressed at some point.

chhagedorn commented on test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java:
```diff
@@ -0,0 +1,412 @@
+ * @bug 8323220
+ * @summary Test loop invariant code motion for cmp nodes through reassociation
+ * @library /test/lib /
+ * @run main compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
```
```suggestion
 * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
```

chhagedorn reviewed (commented):
Only some minor comments. Otherwise, looks good to me, too!

caojoshua commented:
Can ignore warning and let the bot squash everything together.

chhagedorn commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -276,16 +274,15 @@ bool IdealLoopTree::is_associative_cmp(Node* n) {
-    BoolNode* boolOut = n->out(i)->isa_Bool();
-    if (boolOut == nullptr || !(boolOut->_test._test == BoolTest::eq ||
-                                boolOut->_test._test == BoolTest::ne)) {
+    BoolNode *bool_out = n->out(i)->isa_Bool();
```
```suggestion
    BoolNode* bool_out = n->out(i)->isa_Bool();
```

caojoshua replied:
Thanks, committed suggestion through GitHub

chhagedorn reviewed (approved):
Thanks for making the changes. One more minor thing, otherwise, looks good!

chhagedorn commented:
> @eme64 could you take another pass at this?

Just to let you know, he will be back next week.

eme64 reviewed (approved):
@chhagedorn for the suggestions, I like the updates!
@caojoshua approved again :)

caojoshua commented:
/integrate

caojoshua commented:
@eme64 Thanks for working through this with me. Could you sponsor this?

eme64 commented:
/sponsor
